Basic implementation of Speech API.

Fematich opened the pull request:
I have made a basic implementation of Speech API.
This currently only supports `syncrecognize` from the REST-API.

I think I followed all the contribution guidelines:
- Documented in both the API and narrative documentation (in docs/).
- Fully working on the following CPython versions: 2.7, 3.4, and 3.5 on both UNIX and Windows.
- Feature added no new dependencies.

However, this is my first contribution, so in case I missed something, please let me know and I will see how to fix it!

tseaver commented on google/cloud/speech/client.py:
```diff
@@ -0,0 +1,168 @@
+
+        if (content is None) and (uri is None):
+            message = 'content and uri cannot be both equal to None'
+            raise ValueError(message)
```
Just pass the literal string to the `ValueError` constructor (here and in the next test).

Fematich replied:
Fixed in commit `a5f9c0b`.

tseaver commented on google/cloud/speech/client.py:
```diff
@@ -0,0 +1,168 @@
+        for param_name, param in required_params:
+            if param is None:
+                message = '%r cannot be None' % (param_name)
+                raise ValueError(message)
```
Rather than looping, just check these with individual `if` statements:  move them up below the other validation tests above.

Fematich replied:
Fixed in commit `a5f9c0b`.

tseaver commented on google/cloud/speech/client.py:
```diff
@@ -0,0 +1,168 @@
+            if param is None:
+                message = '%r cannot be None' % (param_name)
+                raise ValueError(message)
+        config = dict(required_params)
```
Use a dict literal here:

```python
    config = {'encoding': encoding, 'sampleRate': sample_rate}
```

Fematich replied:
Fixed in commit `a5f9c0b`.

tseaver commented on google/cloud/speech/client.py:
```diff
@@ -0,0 +1,168 @@
+                ('maxAlternatives', max_alternatives),
+                ('profanityFilter', profanity_filter)]:
+            if param is not None:
+                config[param_name] = param
```
Use individual `if` statements rather than a loop.

Fematich replied:
Fixed in commit `a5f9c0b`.

tseaver commented on unit_tests/speech/test_client.py:
```diff
@@ -0,0 +1,111 @@
+                                                        sample_rate,
+                                                        max_alternatives=2,
+                                                        speech_context=hints)
+        self.assertEqual(speechrecognition_result[0]["transcript"], 'hello')
```
Please split these two into separate tests (one with the URI, one with the content).  Add assertions to each that the right method was called, and that the expected `audio` and `configuration` mappings were passed as the payload.  For one of the two cases, please demonstrate non-default values for the optional parameters (and verify that they get added correctly to the `configuration` mapping).

Fematich replied:
Only 1 method remains. Extra tests with `configuration`, `audio` and non-default values for optional parameters has been added.

Fematich replied:
Fixed in commit `988b7d6`.

tseaver commented on unit_tests/speech/test_client.py:
```diff
@@ -0,0 +1,111 @@
+        with self.assertRaises(ValueError):
+            client.syncrecognize(None, None, None, None)
+        with self.assertRaises(ValueError):
+            client.syncrecognize(None, "uri", None, None)
```
Please split each failure path out into its own testcase, with a method name which describes it.

Fematich replied:
Fixed in commit `a5f9c0b`.

Fematich commented:
Thank you for your feedback! I have implemented your suggestions, except for the additional examples with extra arguments in the docs. I will try to finish this tomorrow.

Fematich commented on docs/index.rst:
```diff
@@ -230,3 +238,4 @@ Cloud Storage
   bucket = client.get_bucket('<your-bucket-name>')
   blob = bucket.blob('my-test-file.txt')
   blob.upload_from_string('this is test content!')
+
```
Fixed in commit `a5f9c0b`.

Fematich commented on docs/speech-usage.rst:
```diff
@@ -0,0 +1,61 @@
+
+At this moment we only support one method of the Speech API:
+
+- `syncrecognize`_
```
Fixed in commit `a5f9c0b`.

Fematich commented on docs/speech-usage.rst:
```diff
@@ -0,0 +1,61 @@
+
+- `syncrecognize`_
+
+Synchronous Recognize
```
Fixed in commit `a5f9c0b`.

Fematich commented on docs/speech-usage.rst:
```diff
@@ -0,0 +1,61 @@
+
+Synchronous Recognize
+---------------------
+
```
Fixed in commit `a5f9c0b`.

Fematich commented on docs/speech-usage.rst:
```diff
@@ -0,0 +1,61 @@
+---------------------
+
+The :meth:`~google.cloud.speech.Client.syncrecognize` method
+does speech to text on a file and returns the text
```
Fixed in commit `a5f9c0b`.

Fematich commented on docs/speech-usage.rst:
```diff
@@ -0,0 +1,61 @@
+
+The :meth:`~google.cloud.speech.Client.syncrecognize` method
+does speech to text on a file and returns the text
+as a :class:`list` of tuples dicts (each containing a transcript an a confidence value).
```
Fixed in commit `a5f9c0b`.

Fematich commented on google/cloud/speech/client.py:
```diff
@@ -0,0 +1,168 @@
+
+        See `syncrecognize`_.
+
+        :type content: str
```
Fixed in commit `a5f9c0b`.

Fematich commented on google/cloud/speech/client.py:
```diff
@@ -0,0 +1,168 @@
+                      data bytes encoded as specified in RecognitionConfig.
+                      This is a base64-encoded string
+
+        :type uri: str
```
As you suggested in the previous suggestion, I removed the support for uri/source_uri at this stage.

Fematich replied:
Fixed in commit `988b7d6`.

Fematich commented on google/cloud/speech/client.py:
```diff
@@ -0,0 +1,168 @@
+                              words to the vocabulary of the recognizer.
+
+        :rtype: list
+        :returns: A list of tuples. One tuple for each alternative. Each tuple
```
Fixed in commit `a5f9c0b`.

Fematich commented on docs/speech-client.rst:
```diff
@@ -0,0 +1,16 @@
+  :members:
+  :undoc-members:
+  :show-inheritance:
+
```
Fixed in commit `a5f9c0b`.

tseaver commented on google/cloud/speech/client.py:
```diff
@@ -60,28 +62,19 @@ class Client(client_module.Client):
-                      supported, which must be specified in the following
-                      format: gs://bucket_name/object_name
+        :type content: bytes
+        :param content: Byte stream of audio.
```
Why drop the `source_uri` support?  I don't think I asked for that (I did ask to have it renamed from `uri` -> `source_uri`).

Fematich replied:
I misunderstood this comment:

>  ISTM to be more useful to accept (only) bytes, and then perform the base64 encoding on them (see the Vision Image constructor).

I will add it back in the next commit.

Fematich replied:
Fixed in commit `988b7d6`.

daspecster commented on google/cloud/speech/client.py:
```diff
@@ -0,0 +1,157 @@
+    """Client to bundle configuration needed for API requests.
+
+    :type project: str
+    :param project: the project which the client acts on behalf of. Will be
```
"the" -> "The".
Also, seems like "on behalf of. Will be" should be "on behalf of, will be".?

dhermes replied:
I'd guess this docstring is copy-pasted from elsewhere. If so, it's fine to leave as-is.

Fematich replied:
I did indeed copy-paste it from somewhere else, but it's updated now.

daspecster commented on google/cloud/speech/client.py:
```diff
@@ -0,0 +1,157 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Basic client for Google Cloud Speech API."""
```
I think there should be a newline after the module docstring.

dhermes replied:
Yup. I think you meant a blank line, right?

Fematich replied:
Done :-)!

dhermes commented on docs/index.rst:
```diff
@@ -229,4 +237,4 @@ Cloud Storage
   bucket = client.get_bucket('<your-bucket-name>')
   blob = bucket.blob('my-test-file.txt')
-  blob.upload_from_string('this is test content!')
+  blob.upload_from_string('this is test content!')
```
I'm wondering why this shows up in the diff?

Fematich replied:
Very strange indeed :-)!

dhermes replied:
Can you track down what the issue might be?

tseaver replied:
Is there trailing whitespace somewhere?

dhermes replied:
Or a mix of Windows and Unix line endings?

dhermes replied:
I checked out the branch. Confirmed it was trailing whitespace (or absence of it).

dhermes commented on docs/speech-usage.rst:
```diff
@@ -0,0 +1,54 @@
+  .. code-block:: python
+
+     >>> alternatives = client.sync_recognize(None,"gs://my-bucket/recording.flac",
+     ...                 "FLAC", 16000, max_alternatives=2):
```
Please reformat:

```python
     >>> alternatives = client.sync_recognize(
     ...     None, 'gs://my-bucket/recording.flac',
     ...     'FLAC', 16000, max_alternatives=2)
```

dhermes commented on docs/speech-usage.rst:
```diff
@@ -0,0 +1,54 @@
+     >>> for alternative in alternatives:
+     ...     print('=' * 20)
+     ...     print('   transcript: %s' % (alternative["transcript"],))
+     ...     print('   confidence: %s' % (alternative["confidence"],))
```
Can you use single quotes, e.g. `'confidence'` not `"confidence"`

Fematich replied:
Fixed in commit `988b7d6`.

dhermes commented on docs/speech-usage.rst:
```diff
@@ -0,0 +1,54 @@
+     ...     print('   confidence: %s' % (alternative["confidence"],))
+     ====================
+              transcript: Hello, this is a test
+              confidence: 0.81
```
This indent doesn't match the above, also, not needed?

dhermes commented on google/cloud/speech/__init__.py:
```diff
@@ -0,0 +1,18 @@
+
+"""Google Cloud Speech API wrapper."""
+
+from google.cloud.speech.client import Client, Encoding
```
Please refrain from commas in import. Just turn this into two lines

```python
from google.cloud.speech.client import Client
from google.cloud.speech.client import Encoding
```

Fematich replied:
Fixed in commit `988b7d6`.

dhermes commented on google/cloud/speech/client.py:
```diff
@@ -0,0 +1,157 @@
+
+
+class Encoding(object):
+    """Audio encoding types."""
```
Can you provide a link to the API docs for these?

Fematich replied:
Fixed in commit `988b7d6`.

dhermes commented on google/cloud/speech/client.py:
```diff
@@ -0,0 +1,157 @@
+        """Synchronous Speech Recognition.
+
+        .. _sync_recognize: https://cloud.google.com/speech/reference/\
+                             rest/v1beta1/speech/syncrecognize
```
This (`rest/...`) should line up with the text above (`https...`)

Fematich replied:
I changed it, because of PEP8 issues, only other option is to put the whole url on the line below?

dhermes replied:
Multiple lines is fine. Just make the two parts line up vertically.

dhermes commented on google/cloud/speech/client.py:
```diff
@@ -0,0 +1,157 @@
+
+        :type sample_rate: int
+        :param sample_rate: Sample rate in Hertz of the audio data sent in all
+                           RecognitionAudio messages. Valid values are: 8000-
```
1. This indent doesn't line up with the text above. 
1. `8000-48000` should be together on a single line
1. What is `RecognitionAudio`?

Fematich replied:
Fixed in commit `988b7d6`.

dhermes commented on google/cloud/speech/client.py:
```diff
@@ -0,0 +1,157 @@
+
+        :type language_code: str
+        :param language_code: (Optional) The language of the supplied audio as
+                             BCP-47 language tag. Example: "en-GB".
```
Ditto here on the indent

Fematich replied:
Fixed in commit `988b7d6`.

dhermes commented on google/cloud/speech/client.py:
```diff
@@ -0,0 +1,157 @@
+
+        :type max_alternatives: int
+        :param max_alternatives: (Optional) Maximum number of recognition
+                                hypotheses to be returned. The server may
```
Ditto on indent

Fematich replied:
Fixed in commit `988b7d6`.

dhermes commented on google/cloud/speech/client.py:
```diff
@@ -0,0 +1,157 @@
+
+        :type profanity_filter: bool
+        :param profanity_filter: If True, the server will attempt to filter
+                                out profanities, replacing all but the
```
Ditto on indent

Fematich replied:
Fixed in commit `988b7d6`.

dhermes commented on google/cloud/speech/client.py:
```diff
@@ -0,0 +1,157 @@
+        :type profanity_filter: bool
+        :param profanity_filter: If True, the server will attempt to filter
+                                out profanities, replacing all but the
+                                initialcharacter in each filtered word with
```
"initialcharacter" -> "initial character"

Fematich replied:
Fixed in commit `988b7d6`.

dhermes commented on google/cloud/speech/client.py:
```diff
@@ -0,0 +1,157 @@
+
+        :type speech_context: list
+        :param speech_context: A list of strings (max 50) containing words and
+                              phrases "hints" so that the speech recognition
```
Ditto on indent

Fematich replied:
Fixed in commit `988b7d6`.

dhermes commented on google/cloud/speech/client.py:
```diff
@@ -0,0 +1,157 @@
+        if encoding is None:
+            raise ValueError('encoding cannot be None')
+        if sample_rate is None:
+            raise ValueError('sample_rate cannot be None')
```
Please be consistent using single and double quotes. I prefer single quotes but @tseaver uses double quotes in these situations (i.e. situations with text intended to be read by a human)

Fematich replied:
Fixed in commit `988b7d6`. I will stick to single quotes for consistency.

dhermes commented on google/cloud/speech/client.py:
```diff
@@ -0,0 +1,157 @@
+        if profanity_filter is not None:
+            config['profanityFilter'] = profanity_filter
+        if speech_context is not None:
+            config["speechContext"] = {"phrases": speech_context}
```
Single quotes please

Fematich replied:
Fixed in commit `988b7d6`.

dhermes commented on google/cloud/speech/client.py:
```diff
@@ -0,0 +1,157 @@
+
+        data = {
+            'audio': audio,
+            'config': config
```
Please add trailing comma: `'config': config,`

Fematich replied:
Fixed in commit `988b7d6`.

dhermes commented on google/cloud/speech/client.py:
```diff
@@ -0,0 +1,157 @@
+        api_response = self.connection.api_request(
+            method='POST', path='syncrecognize', data=data)
+
+        return api_response['results'][0]['alternatives']
```
Please break this up into intermediate values to make it clear what is happening. ISTM you are making an assumption that `results` will always be length 1. The code should verify this.

Fematich replied:
Fixed in commit `988b7d6`.

dhermes commented on google/cloud/speech/client.py:
```diff
@@ -0,0 +1,157 @@
+        }
+
+        api_response = self.connection.api_request(
+            method='POST', path='syncrecognize', data=data)
```
Does the API support `fields`? ISTM we could make a smaller payload since `alternatives` is all you want from the resulting payload.

Fematich replied:
I tested the fields parameter [here](https://cloud.google.com/speech/reference/rest/v1beta1/speech/syncrecognize). You can give a fields parameter, but the only possible output is `alternatives`. The only way to get a smaller payload is to ignore the `confidence` value in the `alternatives` dict. But this didn't seem like a good idea.

dhermes replied:
Got it. I see now that this is the entire body.

dhermes commented on scripts/verify_included_modules.py:
```diff
@@ -55,6 +55,7 @@
     'google.cloud.translate.__init__',
     'google.cloud.vision.__init__',
     'google.cloud.vision.fixtures',
+    'google.cloud.speech.__init__',
```
Please put this in alphabetical order.

dhermes commented on unit_tests/speech/_fixtures.py:
```diff
@@ -0,0 +1,12 @@
+SYNC_RECOGNIZE_RESPONSE = {
```
1. Can you put the `Copyright` header in this file as well?
1. Is this file really needed? This variable is tiny and could just as easily be put inline wherever it is used.

Fematich replied:
1. Done
2. I also put it in there, because the number of variables will include if we add more functionalities (e.g. for StreamingRecognition)

dhermes commented on unit_tests/speech/test_client.py:
```diff
@@ -0,0 +1,165 @@
+from google.cloud._helpers import _to_bytes
+
+_AUDIO_CONTENT = _to_bytes('/9j/4QNURXhpZgAASUkq')
+_B64_AUDIO_CONTENT = base64.b64encode(_AUDIO_CONTENT)
```
Please move all non-`unittest` imports into test cases and helpers. See the [blog post][1] by @tseaver on this philosophy.

[1]: https://palladion.com//home/tseaver/obzervationz/2008/unit_testing_notes-20080724

Fematich replied:
Thanks for the link, moved!

dhermes commented on unit_tests/speech/test_client.py:
```diff
@@ -0,0 +1,165 @@
+
+class TestClient(unittest.TestCase):
+    SAMPLE_RATE = 16000
+    HINTS = ["hi"]
```
Single quotes

Fematich replied:
Fixed in commit `988b7d6`.

dhermes commented on unit_tests/speech/test_client.py:
```diff
@@ -0,0 +1,165 @@
+        creds = _Credentials()
+        http = object()
+        client = self._makeOne(credentials=creds, http=http)
+        self.assertTrue(isinstance(client.connection, Connection))
```
You could / should use `self.assertIsInstance`. This method was absent in Python 2.6, which we used to support, which is why you find it sprinkled in some of our tests.

Fematich replied:
Fixed in commit `988b7d6`.

dhermes commented on unit_tests/speech/test_client.py:
```diff
@@ -0,0 +1,165 @@
+            },
+            "audio": {
+                "content": _B64_AUDIO_CONTENT
+            }
```
Single quotes

Fematich replied:
Fixed in commit `988b7d6`.

dhermes commented on unit_tests/speech/test_client.py:
```diff
@@ -0,0 +1,165 @@
+        response = client.sync_recognize(_AUDIO_CONTENT,
+                                         encoding,
+                                         self.SAMPLE_RATE,
+                                         language_code="EN",
```
Single quotes

Fematich replied:
Fixed in commit `988b7d6`.

dhermes commented on unit_tests/speech/test_client.py:
```diff
@@ -0,0 +1,165 @@
+
+        self.assertEqual(REQUEST,
+                         client.connection._requested[0]['data'])
+        self.assertEqual(response[0]["transcript"], 'hello')
```
1. Single quotes
1. Seems these assertions are not quite enough. Would like to see more comprehensive checking of the contents of  `client.connection._requested` and `response`

dhermes commented on unit_tests/speech/test_client.py:
```diff
@@ -0,0 +1,165 @@
+            },
+            "audio": {
+                "content": _B64_AUDIO_CONTENT
+            }
```
Single quotes and trailing commas please

Fematich replied:
Fixed in commit `988b7d6`.

dhermes commented on unit_tests/speech/test_client.py:
```diff
@@ -0,0 +1,165 @@
+
+        self.assertEqual(REQUEST,
+                         client.connection._requested[0]['data'])
+        self.assertEqual(response[0]["transcript"], 'hello')
```
Ditto here as in the previous test case

Fematich replied:
Fixed in commit `988b7d6`.

dhermes commented on unit_tests/speech/test_client.py:
```diff
@@ -0,0 +1,165 @@
+
+class _Credentials(object):
+
+    _scopes = ('https://www.googleapis.com/auth/cloud-platform')
```
You're missing a comma in `('https://...')` to make it a tuple. Right now the parens don't do anything.

Fematich replied:
Fixed in commit `988b7d6`.

dhermes commented:
@Fematich Really great work! Especially impressed that you have Travis 100% green on your first try!

Fematich commented:
I have updated the code (included source_uri again and updated formatting/unit_tests).
I still have to expand speech-usage.rst and the  more comprehensive checking of the contents of `client.connection._requested` and `response`

dhermes commented on google/cloud/speech/client.py:
```diff
@@ -0,0 +1,182 @@
+
+        :type source_uri: str
+        :param source_uri: URI that points to a file that contains audio
+                      data bytes as specified in RecognitionConfig.
```
This text should be lines up with `URI`. Or, if you want more space, move all the text below and indent 4 spaces.

```
        :param source_uri: 
            URI that points to a file that contains audio
            data ...
```

dhermes commented on google/cloud/speech/client.py:
```diff
@@ -0,0 +1,182 @@
+                    between 0 and 1.
+        """
+
+        if (content is None) and (source_uri is None):
```
The parentheses are superflous. Please drop them

```python
if content is None and source_uri is None:
```

dhermes commented on google/cloud/speech/client.py:
```diff
@@ -0,0 +1,182 @@
+            raise ValueError('content and source_uri cannot be both equal to\
+                             None')
+
+        if (content is not None) and (source_uri is not None):
```
Ditto on parens

dhermes commented on google/cloud/speech/client.py:
```diff
@@ -0,0 +1,182 @@
+        """
+
+        if (content is None) and (source_uri is None):
+            raise ValueError('content and source_uri cannot be both equal to\
```
Please don't use the backlash continuation. Instead use implict string concat:

```python
            raise ValueError('content and source_uri cannot be both equal '
                             'to None')
```

dhermes commented on google/cloud/speech/client.py:
```diff
@@ -0,0 +1,182 @@
+                             None')
+
+        if (content is not None) and (source_uri is not None):
+            raise ValueError('content and source_uri cannot be both different from\
```
Ditto on backslash continuation

dhermes commented on google/cloud/speech/client.py:
```diff
@@ -0,0 +1,182 @@
+        if sample_rate is None:
+            raise ValueError('sample_rate cannot be None')
+
+        if content is not None:
```
`content` is a required argument but you're treating it as optional here. We somewhat covered this pattern in [`google.cloud.language.client`][1] and [`google.cloud.language.document`][2]

[1]: https://github.com/GoogleCloudPlatform/google-cloud-python/blob/d149be50977903ace57a0e3a998b6fb6b4c74f93/google/cloud/language/client.py
[2]: https://github.com/GoogleCloudPlatform/google-cloud-python/blob/d149be50977903ace57a0e3a998b6fb6b4c74f93/google/cloud/language/document.py

tseaver replied:
Exactly one of `content` or `source_uri` is required.

dhermes replied:
@tseaver When I say "required" in the above, I mean "positional arg"

tseaver replied:
OK, I see.  ISTM that both should default to `None`, since we have the checks in place that exactly one is supplied.

dhermes replied:
Agreed.

dhermes commented on unit_tests/speech/_fixtures.py:
```diff
@@ -0,0 +1,30 @@
+}
+
+SYNC_RECOGNIZE_EMPTY_RESPONSE = {
+    'results': []
```
Please add trailing comma

dhermes commented on unit_tests/speech/test_client.py:
```diff
@@ -0,0 +1,191 @@
+
+import unittest
+
+_AUDIO_SOURCE_URI = 'gs://sample-bucket/sample-recording.flac'
```
Any reason this can't be a class-level constant?

tseaver commented:
@Fematich Thanks for addressing my review issues.  I'll let @dhermes finish the review and merge.  Thanks again for your work!

dhermes commented:
I'm going to merge as-is. I'll send a PR addressing most of my concerns. Remaining issues outside of that PR will be:

- https://github.com/GoogleCloudPlatform/google-cloud-python/pull/2344/files#r79738759
- https://github.com/GoogleCloudPlatform/google-cloud-python/pull/2344/files#r79307666

dhermes commented:
@Fematich Thanks for your great work! We decided to go ahead and merge to speed up the process.

We'd like to get this library feature-complete ASAP so we didn't want the review cycle to slow us down on getting the first chunks in. From here, a person working on this library full-time will take it to the finish line. You were great!

Fematich commented:
Thank you it was a pleasure :-)!
